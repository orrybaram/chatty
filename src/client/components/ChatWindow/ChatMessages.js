import React, { Component } from 'react';
import { observer, PropTypes } from 'mobx-react';
import shortId from 'shortid';
import classNames from 'classnames';
import ChatMessage from './ChatMessage';

const propTypes = {
  messages: PropTypes.arrayOrObservableArray,
  recipients: PropTypes.arrayOrObservableArray,
  user: PropTypes.objectOrObservableObject
};

class ChatMessages extends Component {
  constructor() {
    super();
    this.messagesWindow = null;
    this.isScrolledToBottom = false;
  }
  componentDidMount() {
    // Kind of a hack, the dom isn't fully painted when this is called
    setTimeout(() => {
      this.scrollToBottom();
    }, 250);
  }
  componentWillUpdate() {
    this.isScrolledToBottom = (
      this.messagesWindow.scrollTop === (
        this.messagesWindow.scrollHeight - this.messagesWindow.offsetHeight
      )
    );
  }
  componentDidUpdate() {
    if (this.isScrolledToBottom) {
      this.scrollToBottom();
    }
  }
  scrollToBottom() {
    this.messagesWindow.scrollTop = this.messagesWindow.scrollHeight;
  }
  render() {
    const { messages, recipients, user } = this.props;
    const messageEls = messages.map((msg) => {
      const isUser = user.id === msg.sentBy.id;
      return (
        <ChatMessage
          key={shortId.generate()}
          user={user}
          isUser={isUser}
          message={msg}
        />
      );
    });

    const typingIndicators = recipients.map((_user) => {
      const classes = classNames([
        'typing-indicator',
        { 'typing-indicator--shown': _user.isTyping }
      ]);
      return (<div key={shortId.generate()} className={classes}>{_user.name} is typing</div>);
    });

    return (
      <div className='chat-window__messages' ref={(ref) => { this.messagesWindow = ref; }}>
        {messageEls}
        <div className='chat-window__messages__typing-indicators'>
          {typingIndicators}
        </div>
      </div>
    );
  }
}

ChatMessages.propTypes = propTypes;
export default observer(ChatMessages);
