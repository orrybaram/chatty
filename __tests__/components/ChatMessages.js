/* eslint-disable import/no-extraneous-dependencies */
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import ChatMessages from '../../src/client/components/ChatWindow/ChatMessages';

describe('ChatMessages', () => {
  it('renders correctly', () => {
    const messages = [{
      body: 'yo',
      dateCreated: new Date(),
      sentBy: { name: 'gary' }
    }, {
      body: 'sup',
      dateCreated: new Date(),
      sentBy: { name: 'bags' }
    }];
    const component = renderer.create(
      <ChatMessages
        user={{}}
        recipients={[]}
        messages={messages}
      />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('shows a typing indicator when a recipient is typing', () => {
    const recipients = [{
      name: 'gary',
      isTyping: false
    }];
    const component = mount(<ChatMessages recipients={recipients} messages={[]} />);
    expect(component.find('.typing-indicator').get(0).className).toBe('typing-indicator');
    component.setProps({ recipients: [{
      name: 'gary',
      isTyping: true
    }] });
    expect(component.find('.typing-indicator').get(0).className).toBe('typing-indicator typing-indicator--shown');
  });
});
