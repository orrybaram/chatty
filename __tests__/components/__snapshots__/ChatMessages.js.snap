exports[`ChatMessages renders correctly 1`] = `
<div
  className="chat-window__messages">
  <div
    className="chat-window__messages__message chat-window__messages__message--user">
    <div
      className="chat-window__messages__message-wrapper">
      <div
        className="chat-window__messages__message__body">
        yo
      </div>
      <div
        className="chat-window__messages__message__date">
        a few seconds ago
      </div>
    </div>
  </div>
  <div
    className="chat-window__messages__message chat-window__messages__message--user">
    <div
      className="chat-window__messages__message-wrapper">
      <div
        className="chat-window__messages__message__body">
        sup
      </div>
      <div
        className="chat-window__messages__message__date">
        a few seconds ago
      </div>
    </div>
  </div>
  <div
    className="chat-window__messages__typing-indicators" />
</div>
`;
